README.md
Django-CMS Pagedown Plugin
===============

A django-cms plugin to allow the use of [Stack Overflow&#39;s &quot;Pagedown&quot; Markdown editor](http://code.google.com/p/pagedown/)

#### Installation ####

- Install via pip: `pip install -e https://github.com/timmyomahony/cmsplugin-pagedown.git#egg=cmsplugin-pagedown` or `pip install cmsplugin-pagedown`
- Add `pagedown` and `cmsplugin_pagedown` to your `INSTALLED_APPS`

#### Usage ####

Simply select `Pagedown Markdown` from the `Available Plugins` dropdown when using a django-cms placeholder. 

#####Screenshot: Example Frontend Output#####

------

![Screenshot of example frontend output](https://github.com/timmyomahony/cmsplugin-pagedown/blob/master/frontend-screenshot.png?raw=true "Screenshot of example frontend output")

#####Screenshot: Plugin#####

------

![Screenshot of a CMS placeholder using pagedown](https://github.com/timmyomahony/cmsplugin-pagedown/blob/master/backend-screenshot.png?raw=true "Screenshot of a CMS placeholder using pagedown")


CHANGES.txt
v0.0.1, 02/05/2012 -- Initial release.
v0.0.2, 27/10/2012 -- Added some CSS


cmsplugin_pagedown/models.py
from django.db import models
from django.template.defaultfilters import striptags
from cms.models import CMSPlugin
import markdown


class PagedownConfig(CMSPlugin):
    markdown = models.TextField()
    html = models.TextField()
    plaintext = models.TextField()

    css_id = models.CharField(max_length=50, blank=True, help_text="Add an optional CSS ID to the rendered outputs wrapping element")
    css_classes = models.CharField(max_length=50, blank=True, help_text="Add CSS classes to the rendered outputs wrapper element")

    def save(self, *args, **kwargs):
        self.html = markdown.markdown(self.markdown)
        self.plaintext = striptags(self.html)
        super(PagedownConfig, self).save(args, kwargs)

    def __unicode__(self):
        return self.plaintext[:50]

    @property
    def wordcount(self):
        return len(self.plaintext)


cmsplugin_pagedown/migrations/0001_initial.py
# -*- coding: utf-8 -*-
import datetime
from south.db import db
from south.v2 import SchemaMigration
from django.db import models


class Migration(SchemaMigration):

    def forwards(self, orm):
        # Adding model 'PagedownConfig'
        db.create_table('cmsplugin_pagedownconfig', (
            ('cmsplugin_ptr', self.gf('django.db.models.fields.related.OneToOneField')(to=orm['cms.CMSPlugin'], unique=True, primary_key=True)),
            ('markdown', self.gf('django.db.models.fields.TextField')()),
            ('html', self.gf('django.db.models.fields.TextField')()),
            ('plaintext', self.gf('django.db.models.fields.TextField')()),
        ))
        db.send_create_signal('cmsplugin_pagedown', ['PagedownConfig'])

    def backwards(self, orm):
        # Deleting model 'PagedownConfig'
        db.delete_table('cmsplugin_pagedownconfig')

    models = {
        'cms.cmsplugin': {
            'Meta': {'object_name': 'CMSPlugin'},
            'creation_date': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'language': ('django.db.models.fields.CharField', [], {'max_length': '15', 'db_index': 'True'}),
            'level': ('django.db.models.fields.PositiveIntegerField', [], {'db_index': 'True'}),
            'lft': ('django.db.models.fields.PositiveIntegerField', [], {'db_index': 'True'}),
            'parent': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['cms.CMSPlugin']", 'null': 'True', 'blank': 'True'}),
            'placeholder': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['cms.Placeholder']", 'null': 'True'}),
            'plugin_type': ('django.db.models.fields.CharField', [], {'max_length': '50', 'db_index': 'True'}),
            'position': ('django.db.models.fields.PositiveSmallIntegerField', [], {'null': 'True', 'blank': 'True'}),
            'rght': ('django.db.models.fields.PositiveIntegerField', [], {'db_index': 'True'}),
            'tree_id': ('django.db.models.fields.PositiveIntegerField', [], {'db_index': 'True'})
        },
        'cms.placeholder': {
            'Meta': {'object_name': 'Placeholder'},
            'default_width': ('django.db.models.fields.PositiveSmallIntegerField', [], {'null': 'True'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'slot': ('django.db.models.fields.CharField', [], {'max_length': '50', 'db_index': 'True'})
        },
        'cmsplugin_pagedown.pagedownconfig': {
            'Meta': {'object_name': 'PagedownConfig', 'db_table': "'cmsplugin_pagedownconfig'", '_ormbases': ['cms.CMSPlugin']},
            'cmsplugin_ptr': ('django.db.models.fields.related.OneToOneField', [], {'to': "orm['cms.CMSPlugin']", 'unique': 'True', 'primary_key': 'True'}),
            'html': ('django.db.models.fields.TextField', [], {}),
            'markdown': ('django.db.models.fields.TextField', [], {}),
            'plaintext': ('django.db.models.fields.TextField', [], {})
        }
    }

    complete_apps = ['cmsplugin_pagedown']

cmsplugin_pagedown/migrations/0002_auto__add_field_pagedownconfig_css_id__add_field_pagedownconfig_css_cl.py
# -*- coding: utf-8 -*-
import datetime
from south.db import db
from south.v2 import SchemaMigration
from django.db import models


class Migration(SchemaMigration):

    def forwards(self, orm):
        # Adding field 'PagedownConfig.css_id'
        db.add_column('cmsplugin_pagedownconfig', 'css_id',
                      self.gf('django.db.models.fields.CharField')(default='', max_length=50, blank=True),
                      keep_default=False)

        # Adding field 'PagedownConfig.css_classes'
        db.add_column('cmsplugin_pagedownconfig', 'css_classes',
                      self.gf('django.db.models.fields.CharField')(default='', max_length=50, blank=True),
                      keep_default=False)


    def backwards(self, orm):
        # Deleting field 'PagedownConfig.css_id'
        db.delete_column('cmsplugin_pagedownconfig', 'css_id')

        # Deleting field 'PagedownConfig.css_classes'
        db.delete_column('cmsplugin_pagedownconfig', 'css_classes')


    models = {
        'cms.cmsplugin': {
            'Meta': {'object_name': 'CMSPlugin'},
            'changed_date': ('django.db.models.fields.DateTimeField', [], {'auto_now': 'True', 'blank': 'True'}),
            'creation_date': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'language': ('django.db.models.fields.CharField', [], {'max_length': '15', 'db_index': 'True'}),
            'level': ('django.db.models.fields.PositiveIntegerField', [], {'db_index': 'True'}),
            'lft': ('django.db.models.fields.PositiveIntegerField', [], {'db_index': 'True'}),
            'parent': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['cms.CMSPlugin']", 'null': 'True', 'blank': 'True'}),
            'placeholder': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['cms.Placeholder']", 'null': 'True'}),
            'plugin_type': ('django.db.models.fields.CharField', [], {'max_length': '50', 'db_index': 'True'}),
            'position': ('django.db.models.fields.PositiveSmallIntegerField', [], {'null': 'True', 'blank': 'True'}),
            'rght': ('django.db.models.fields.PositiveIntegerField', [], {'db_index': 'True'}),
            'tree_id': ('django.db.models.fields.PositiveIntegerField', [], {'db_index': 'True'})
        },
        'cms.placeholder': {
            'Meta': {'object_name': 'Placeholder'},
            'default_width': ('django.db.models.fields.PositiveSmallIntegerField', [], {'null': 'True'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'slot': ('django.db.models.fields.CharField', [], {'max_length': '50', 'db_index': 'True'})
        },
        'cmsplugin_pagedown.pagedownconfig': {
            'Meta': {'object_name': 'PagedownConfig', 'db_table': "'cmsplugin_pagedownconfig'", '_ormbases': ['cms.CMSPlugin']},
            'cmsplugin_ptr': ('django.db.models.fields.related.OneToOneField', [], {'to': "orm['cms.CMSPlugin']", 'unique': 'True', 'primary_key': 'True'}),
            'css_classes': ('django.db.models.fields.CharField', [], {'max_length': '50', 'blank': 'True'}),
            'css_id': ('django.db.models.fields.CharField', [], {'max_length': '50', 'blank': 'True'}),
            'html': ('django.db.models.fields.TextField', [], {}),
            'markdown': ('django.db.models.fields.TextField', [], {}),
            'plaintext': ('django.db.models.fields.TextField', [], {})
        }
    }

    complete_apps = ['cmsplugin_pagedown']